src/librbd/cache/rwl/Request.cc

```diff
@@ -0,0 +1,557 @@
+  }
+
+  ceph_assert(guard_ctx.m_cell);
+  this->detained = guard_ctx.m_state.detained; /* overlapped */
+  this->m_queued = guard_ctx.m_state.queued; /* queued behind at least one barrier */
+  this->set_cell(guard_ctx.m_cell);
+}
+
+template <typename T>
+void C_WriteRequest<T>::finish_req(int r) {
+  if (RWL_VERBOSE_LOGGING) {
+    ldout(rwl.m_image_ctx.cct, 15) << "write_req=" << this << " cell=" << this->get_cell() << dendl;
+  }
+
+  /* Completed to caller by here (in finish(), which calls this) */
+  utime_t now = ceph_clock_now();
+  rwl.release_write_lanes(this);
+  this->release_cell(); /* TODO: Consider doing this in appending state */
+  update_req_stats(now);
+}
+
+template <typename T>
+void C_WriteRequest<T>::setup_buffer_resources(uint64_t &bytes_cached, uint64_t &bytes_dirtied) {
+  for (auto &extent : this->image_extents) {
+    resources.buffers.emplace_back();
+    struct WriteBufferAllocation &buffer = resources.buffers.back();
+    buffer.allocation_size = MIN_WRITE_ALLOC_SIZE;
+    buffer.allocated = false;
+    bytes_cached += extent.second;
+    if (extent.second > buffer.allocation_size) {
+      buffer.allocation_size = extent.second;
+    }
+  }
+  bytes_dirtied = bytes_cached;
+}
+
+template <typename T>
+void C_WriteRequest<T>::setup_log_operations() {
+  for (auto &extent : this->image_extents) {
+    /* operation->on_write_persist connected to m_prior_log_entries_persisted Gather */
+    auto operation =
+      std::make_shared<WriteLogOperation<T>>(*m_op_set, extent.first, extent.second);
+    m_op_set->operations.emplace_back(operation);
+  }
+}
+
+template <typename T>
+void C_WriteRequest<T>::schedule_append() {
+  ceph_assert(++m_appended == 1);
+  if (m_do_early_flush) {
+    /* This caller is waiting for persist, so we'll use their thread to
+     * expedite it */
+    rwl.flush_pmem_buffer(this->m_op_set->operations);
+    rwl.schedule_append(this->m_op_set->operations);
+  } else {
+    /* This is probably not still the caller's thread, so do the payload
+     * flushing/replicating later. */
+    rwl.schedule_flush_and_append(this->m_op_set->operations);
+  }
+}
+
+/**
+ * Attempts to allocate log resources for a write. Returns true if successful.
+ *
+ * Resources include 1 lane per extent, 1 log entry per extent, and the payload
+ * data space for each extent.
+ *
+ * Lanes are released after the write persists via release_write_lanes()
+ */
+template <typename T>
+bool C_WriteRequest<T>::alloc_resources()
+{
+  bool alloc_succeeds = true;
+  bool no_space = false;
+  utime_t alloc_start = ceph_clock_now();
+  uint64_t bytes_allocated = 0;
+  uint64_t bytes_cached = 0;
+  uint64_t bytes_dirtied = 0;
+
+  ceph_assert(!rwl.m_lock.is_locked_by_me());
+  ceph_assert(!resources.allocated);
+  resources.buffers.reserve(this->image_extents.size());
+  {
+    std::lock_guard locker(rwl.m_lock);
+    if (rwl.m_free_lanes < this->image_extents.size()) {
```

Comment:
General delegation of responsibility comment: this class could ask the RWL for the required space and get the references to the spaces back from the RWL -- instead of embedding this logic in each request class.

Reply:
The check is only for write requests which need to allocate data space in the pmem device. It controls the number of concurrently asking for space allocation on the pmem device. C_WriteRequest is the class for normal write requests, also it is the base class of  C_CompAndWriteRequest and C_WriteSameRequest.   So the check is not redundant in other places.

Reply:
We would like to move alloc_resources to the ReplicatedWriteLog class and overloads it with different Requests. eg,
```c++
bool alloc_resources(C_CompAndWriteRequest *req);
bool alloc_resources(C_WriteSameRequest *req);
...
```